torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -622,6 +655,19 @@ def _get_gradient_predivide_factor(self, world_size: int) -> float:
             factor *= 2
         return float(factor)
 
+    def _get_param_to_param_index(self) -> Dict[FlatParameter, int]:
+        """
+        Returns a mapping from FSDP flattened parameter to a unique index in
+        the FSDP module hierarchy rooted at this module. Critically, this
+        mapping must be consistent across ranks.
+        """
+        assert self._is_root, "Expects to be called only on root FSDP modules"
+        param_to_param_index = {
+            cast(FlatParameter, param): i
+            for i, param in enumerate(self.parameters())
```

Comment:
Does `.parameters()` only contain `FlatParameter` in all cases? Is there a reason we can't use `self.params`?

Reply:
`self.params` only contains the root module's parameters. `.parameters()` includes the `FlatParameter`s of all nested FSDP modules as well.